pymc_marketing/mmm/delayed_saturated_mmm.py

```diff
@@ -63,80 +65,170 @@ def __init__(
 
-        else:
-            fourier_features = None
+        model_data_dict = {
+            "channel_data_": {
+                "type": "MutableData",
+                "value": channel_data,
+                "dims": ("date", "channel"),
+            },
+            "target_data_": {
+                "type": "MutableData",
+                "value": target_data,
+                "dims": "date",
+            },
+        }
+        self.model_coords = coords
+
+        if control_data is not None:
+            model_data_dict["control_data"] = {
+                "value": control_data,
+                "dims": ("date", "control"),
+            }
+
+        if fourier_features is not None:
+            model_data_dict["fourier_features"] = {
+                "value": fourier_features,
+                "dims": ("date", "fourier_mode"),
+            }
+        model_data = pd.DataFrame.from_dict(
+            model_data_dict
+        )  # change how DataFrame is created
+        return model_data
+
+    def _preprocess_channel_prior(self) -> TensorVariable:
+        return (
+            pm.HalfNormal.dist(sigma=2, shape=len(self.channel_columns))
+            if self.channel_prior is None
+            else change_dist_size(
+                dist=self.channel_prior, new_size=len(self.channel_columns)
+            )
+        )
+
+    def build_model(
+        self,
+        data: Optional[Union[np.ndarray, pd.DataFrame, pd.Series]] = None,
+        model_config: Optional[Dict] = None,
+        **kwargs,
+    ) -> None:
+        self.output_var = "target"
+        self.data = self.generate_model_data(data=data)
 
-        with pm.Model(coords=coords) as self.model:
+        if not model_config:
+            model_config = self.default_model_config
+        with pm.Model(coords=self.model_coords) as self.model:
             channel_data_ = pm.MutableData(
                 name="channel_data",
-                value=channel_data,
-                dims=("date", "channel"),
+                value=self.data.channel_data_.value,
+                dims=self.data.channel_data_.dims,
             )
 
-            target_ = pm.MutableData(name="target", value=target_data, dims="date")
+            target_ = pm.MutableData(
+                name="target",
+                value=self.data["target_data_"]["value"],
+                dims=self.data["target_data_"]["dims"],
+            )
 
-            intercept = pm.Normal(name="intercept", mu=0, sigma=2)
+            intercept = pm.Normal(
+                name="intercept",
+                mu=model_config["intercept"]["mu"],
+                sigma=model_config["intercept"]["sigma"],
+            )
 
+            """beta_channel = pm.HalfNormal(
+                name="beta_channel",
+                sigma=model_config["beta_channel"]["sigma"],
+                dims=model_config["beta_channel"]["dims"],
+            )  # ? Allow prior depend on channel costs?"""
```

Comment:
why is this commented?

Reply:
This is actually something I wanted to ask about, but forgot. After you've added the line:
            beta_channel = self.model.register_rv(
                rv_var=channel_prior, name="beta_channel", dims="channel"
            )
under the commented part, tests are failing with:
ValueError: Variable name beta_channel already exist
it does indeed seem like we're trying to register a rv with exactly the same name as the pm.HalfNormal from previous beta_channel, I'm not sure if it's a bug and it should be allowed to do it like that, or it's a typo and it was supposed to be named for example "beta_channel_" and the _ was lost on the way